### ISSUE TITLE
Incorrect function of regex in Perl 5.8.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6201]
Migrated from [rt.perl.org#19767](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=19767) (status was 'resolved')

Searchable as RT19767$

**Comment by p5pRT** [comment]
### From ernest@virtualitas.net

#### Created by ernest@virtualitas.net



Hi\,

it might be\, that I encountered a problem in the regex machine of Perl
5\.8\.0\.

Due to some queries about incorrect functions by the users of my below
mentioned ProgramV I boiled it down to a reproduceable test\.

Please download the small test script from

http&#8203;://www\.virtualitas\.net/perl/aiml/test\.pl\.txt  \(2 KB\)

While it is working correct on Perl 5\.6\.0 in all cases\, Perl 5\.8\.0
yields

> Guessing start of match\, REx \`^\(?\-xism&#8203;:\(?\-xism&#8203;:\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]\)\+?\)\(\\\-\(\.\.\.' against \`Name\-1'\.\.\.
> Found floating substr \`'$ at offset 6\.\.\.
> Could not match STCLASS\.\.\.
> Match rejected by optimizer

if another regex failed before\.

The problem occurs with ActivePerl 5\.8\.0 build 804 on WinNT as well\.

Ernest

\-\-
&nbsp;              ProgramV \- Alice on Perl \- available at
&nbsp;               http&#8203;://www\.virtualitas\.net/perl/aiml/

&nbsp;      VIRTUALITAS \- Manufacturer of fine OOPPS \- since 1996
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\* VIRTUALITAS Inc\.               \*      http&#8203;://www\.virtualitas\.net  \*
\*                                \*                                  \*
\* European Consultant Office     \*                                  \*
\* Internationales Handelszentrum \*                                  \*
\* Friedrichstraße 95             \*   contact&#8203;:Ernest Lergon          \*
\* 10117 Berlin / Germany         \*    mailto&#8203;:Ernest@&#8203;virtualitas\.net \*
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
&nbsp;      PGP\-Fingerprint 6E6F DC17 A886 342D  D63F 7880 12F5 6BA9
&nbsp;        PGP\-Key http&#8203;://www\.virtualitas\.net/Ernest\_Lergon\.asc

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl v5.8.0:

Configured by root at Thu Dec 19 20:10:27 CET 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.4-4gb, archname=i686-linux
    uname='linux erde 2.4.4-4gb #1 wed may 16 00:37:55 gmt 2001 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010315 (SuSE)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/i686-linux
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/root
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=de_DE@euro
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/local/sbin:/root/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/games/bin:/usr/games:/opt/bin:/opt/gnome/bin:/opt/kde2/bin:/opt/kde/bin:/opt/gnome/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @ysth

On 6 Jan 2003 16&#8203;:17&#8203;:29 \-0000\, Ernest Lergon wrote&#8203;:
>it might be\, that I encountered a problem in the regex machine of Perl
>5\.8\.0\.
>
>Due to some queries about incorrect functions by the users of my below
>mentioned ProgramV I boiled it down to a reproduceable test\.
>
>Please download the small test script from
>
>http&#8203;://www\.virtualitas\.net/perl/aiml/test\.pl\.txt  \(2 KB\)

This was posted 2\+ weeks ago to c\.l\.p\.moderated and assuming the above
url is the same\, I started stripping it down and got this far&#8203;:

use strict;
use warnings;
use utf8;
my $NormalChar        = qr/\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/;
my $NormalWord        = qr/$\{NormalChar\}\+?/;
my $PredNameHyphen    = qr/^$\{NormalWord\}\(\\\-$\{NormalWord\}\)\*?$/;

unless \('Name\-1' =~ /^\(true|false\)$/\)
\{
&nbsp;    print "This test properly failed\\n";
\}

if \( 'Name\-1' =~ $PredNameHyphen \)
\{
&nbsp;    print "And this one properly succeeded\\n";
\}
else
\{
&nbsp;    print "But this one didn't succeed\\n";
\}
\_\_END\_\_

Works on 5\.6\.1 and fails on 5\.8\.0 and bleadperl\.

**Comment by p5pRT** [comment]
### From @ysth

On Mon\, 06 Jan 2003 20&#8203;:08&#8203;:23 \-0800\, sthoenna@&#8203;efn\.org wrote&#8203;:
>I started stripping it down and got this far&#8203;:

This succeeds on perl5\.6\.1 \(with \-Mutf8\) and fails on 5\.8\.0 and bleadperl&#8203;:
"a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";

Can someone who knows something about this take a look?
I'd particularly like to know if there is a workaround\.

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 07 Jan 2003 22&#8203;:14&#8203;:59 \-0800\, sthoenna@&#8203;efn\.org \(Yitzchak Scott\-Thoennes\) said&#8203;:

&nbsp;  > On Mon\, 06 Jan 2003 20&#8203;:08&#8203;:23 \-0800\, sthoenna@&#8203;efn\.org wrote&#8203;:
 >> I started stripping it down and got this far&#8203;:

&nbsp;  > This succeeds on perl5\.6\.1 \(with \-Mutf8\) and fails on 5\.8\.0 and bleadperl&#8203;:
&nbsp;  > "a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";

&nbsp;  > Can someone who knows something about this take a look?
&nbsp;  > I'd particularly like to know if there is a workaround\.

Blame analysis follows\. Note\, that these versions don't compile with
gcc\-3\.2\.1\. I downgraded to 2\.95\.3\.

\-\-\-\-Program\-\-\-\-
use utf8;
"a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";

\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11649/bin/perl\-\-\-\-
ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11650/bin/perl\-\-\-\-
not ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-


HTH\,
\-\- 
andreas

**Comment by p5pRT** [comment]
### From @ysth

On Thu\, 09 Jan 2003 20&#8203;:59&#8203;:36 \+0100\, andreas\.koenig@&#8203;anima\.de wrote&#8203;:
>>>>>> On Tue\, 07 Jan 2003 22&#8203;:14&#8203;:59 \-0800\, sthoenna@&#8203;efn\.org \(Yitzchak Scott\-Thoennes\) said&#8203;:
>
>  > On Mon\, 06 Jan 2003 20&#8203;:08&#8203;:23 \-0800\, sthoenna@&#8203;efn\.org wrote&#8203;:
> >> I started stripping it down and got this far&#8203;:
>
>  > This succeeds on perl5\.6\.1 \(with \-Mutf8\) and fails on 5\.8\.0 and bleadperl&#8203;:
>  > "a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";
>
>  > Can someone who knows something about this take a look?
>  > I'd particularly like to know if there is a workaround\.
>
>Blame analysis follows\. Note\, that these versions don't compile with
>gcc\-3\.2\.1\. I downgraded to 2\.95\.3\.
>
>\-\-\-\-Program\-\-\-\-
>use utf8;
>"a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";
>
>\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11649/bin/perl\-\-\-\-
>ok
>
>\-\-\-\-EOF \($?='0'\)\-\-\-\-
>\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11650/bin/perl\-\-\-\-
>not ok
>
>\-\-\-\-EOF \($?='0'\)\-\-\-\-
>
>
>HTH\,
>\-\- 
>andreas

That would suggest the workaround is to utf8&#8203;::upgrade the data before the
match\.

Thanks\, Andreas\.  Could you see if the original reported problem also
breaks at 11650?  I'm not completely sure I've got exactly the same bug\.

I'd also like a clarification of the intent here\.  If you have
non\-utf8 data matching against e\.g\. \\p\{L\}\, it should assume unicode
code points\, yes?

E\.g\. "\\xe9" =~ /\\p\{L\}/\, even without "use locale" and even under "use
locale" with a C locale \(where 0xe9 is not a letter\)\.

**Comment by p5pRT** [comment]
### From @andk

Bug report 19767 seems to uncover at least three bugs\.

>>>>> On Thu\, 09 Jan 2003 12&#8203;:17&#8203;:16 \-0800\, sthoenna@&#8203;efn\.org \(Yitzchak Scott\-Thoennes\) said&#8203;:

&nbsp;  > On Thu\, 09 Jan 2003 20&#8203;:59&#8203;:36 \+0100\, andreas\.koenig@&#8203;anima\.de wrote&#8203;:
 >>>>>>> On Tue\, 07 Jan 2003 22&#8203;:14&#8203;:59 \-0800\, sthoenna@&#8203;efn\.org \(Yitzchak Scott\-Thoennes\) said&#8203;:
 >> 
 >> > On Mon\, 06 Jan 2003 20&#8203;:08&#8203;:23 \-0800\, sthoenna@&#8203;efn\.org wrote&#8203;:
 >> >> I started stripping it down and got this far&#8203;:
 >> 
 >> > This succeeds on perl5\.6\.1 \(with \-Mutf8\) and fails on 5\.8\.0 and bleadperl&#8203;:
 >> > "a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";
 >> 
 >> > Can someone who knows something about this take a look?
 >> > I'd particularly like to know if there is a workaround\.
 >> 
 >> Blame analysis follows\. Note\, that these versions don't compile with
 >> gcc\-3\.2\.1\. I downgraded to 2\.95\.3\.
 >> 
 >> \-\-\-\-Program\-\-\-\-
 >> use utf8;
 >> "a" =~ m/\[b\]/; print "0" =~ /\\p\{N\}\+\\z/ ? "ok\\n" : "not ok\\n";
 >> 
 >> \-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11649/bin/perl\-\-\-\-
 >> ok
 >> 
 >> \-\-\-\-EOF \($?='0'\)\-\-\-\-
 >> \-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11650/bin/perl\-\-\-\-
 >> not ok
 >> 
 >> \-\-\-\-EOF \($?='0'\)\-\-\-\-
 >> 
 >> 
 >> HTH\,
 >> \-\- 
 >> andreas

&nbsp;  > That would suggest the workaround is to utf8&#8203;::upgrade the data before the
&nbsp;  > match\.

&nbsp;  > Thanks\, Andreas\.  Could you see if the original reported problem also
&nbsp;  > breaks at 11650?  I'm not completely sure I've got exactly the same bug\.

No\. The original problem breaks one patch later \(11651\) and I trimmed
the test script a little&#8203;:

\-\-\-\-Program\-\-\-\-
use utf8;
my $attr = 'Name\-1' ;

my $NormalChar          = qr/\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/;
my $NormalWord          = qr/$\{NormalChar\}\+?/;
my $PredNameHyphen      = qr/^$\{NormalWord\}\(\\\-$\{NormalWord\}\)\*?$/;

$attr =~ /^$/;
if      \( $attr =~ $PredNameHyphen \) \{
&nbsp;        print "ok\\n";
\} else \{
&nbsp;        print "not ok\\n";
\}

\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11650/bin/perl\-\-\-\-
ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of /usr/local/perl\-5\.7\.2@&#8203;11651/bin/perl\-\-\-\-
not ok

\-\-\-\-EOF \($?='0'\)\-\-\-\-


&nbsp;  > I'd also like a clarification of the intent here\.  If you have
&nbsp;  > non\-utf8 data matching against e\.g\. \\p\{L\}\, it should assume unicode
&nbsp;  > code points\, yes?

Assume Unicode code points\, sure\. But if the string isn't in Unicode\,
it must be converted\, right? If the encoding is not known\, I think\,
latin1 is the default\, so we're back to Unicode code points\.

&nbsp;  > E\.g\. "\\xe9" =~ /\\p\{L\}/\, even without "use locale" and even under "use
&nbsp;  > locale" with a C locale \(where 0xe9 is not a letter\)\.

If locale and Unicode do not agree\, I'm not sure who wins\. But in our
two new cases locale doesn't play a role\. Watch the following program
fails with all perls I have&#8203;:

&nbsp;  use strict;
&nbsp;  my @&#8203;res;
&nbsp;  my $a = "x\-x";
&nbsp;  my $NormalChar          = qr/\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/;
&nbsp;  my $NormalWord          = qr/$\{NormalChar\}\+?/;
&nbsp;  my $PredNameHyphen      = qr/^$\{NormalWord\}\(\\\-$\{NormalWord\}\)\*?$/;
&nbsp;  push @&#8203;res\, "ok" if $a =~ $PredNameHyphen;
&nbsp;  $a =~ /^$/;
&nbsp;  push @&#8203;res\, "ok" if $a =~ $PredNameHyphen;
&nbsp;  print @&#8203;res==2 && $res\[0\] eq $res\[1\] ? "ok\\n" : "not ok\\n";

\-\- 
andreas

**Comment by p5pRT** [comment]
### From ernest@virtualitas.net

"Andreas Koenig \(via RT\)" wrote&#8203;:
> 
> If locale and Unicode do not agree\, I'm not sure who wins\. But in our
> two new cases locale doesn't play a role\. Watch the following program
> fails with all perls I have&#8203;:
> 
>   use strict;
>   my @&#8203;res;
>   my $a = "x\-x";
>   my $NormalChar          = qr/\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/;
>   my $NormalWord          = qr/$\{NormalChar\}\+?/;
>   my $PredNameHyphen      = qr/^$\{NormalWord\}\(\\\-$\{NormalWord\}\)\*?$/;
>   push @&#8203;res\, "ok" if $a =~ $PredNameHyphen;
>   $a =~ /^$/;
>   push @&#8203;res\, "ok" if $a =~ $PredNameHyphen;
>   print @&#8203;res==2 && $res\[0\] eq $res\[1\] ? "ok\\n" : "not ok\\n";
> 
Turn on warnings and you'll see the reason for Perl 5\.6\.x \- e\.g\.&#8203;:

Unrecognized escape \\p in character class passed through before HERE
mark in regex m/\[\\p \<\< HERE \{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/ etc\.

After inserting 'use utf8;' it runs fine on Perl 5\.6\.x\.

Another variation\, which might help finding the problem\, is given below\.

Thank you for working on it\!

Ernest

\-\-\-\- CUT \-\-\-\-

use strict;

use warnings;
use utf8;

use Unicode&#8203;::String;

printf "Perl v%vd\\n"\, $^V;

my @&#8203;res;

my $a;

\#    $a = "x\-x";
\#
\#    ok\, but&#8203;:
\#
\#    $a = "x\-ä";
\#
\#    Perl 5\.6\.0 warning&#8203;:
\#    Malformed UTF\-8 character\.
\#
\#    Perl 5\.8\.0 warning
\#    Malformed UTF\-8 character \(1 byte\, need 3\, after start byte 0xe4\)\.
\#
\#    so we have to protect and convert the string constant&#8203;:
\{
&nbsp;    no utf8;

\#    $a = "x\-x";
&nbsp;    $a = "x\-ä";

&nbsp;    print "a ='$a'\\n";

&nbsp;    $a = Unicode&#8203;::String&#8203;::latin1 \( $a \)\->utf8;
\}

print "au='$a'\\n";

my $NormalChar          = qr/\[\\p\{IsDigit\}\\p\{IsLower\}\\p\{IsUpper\}\]/;
my $NormalWord          = qr/$\{NormalChar\}\+?/;
my $PredNameHyphen      = qr/^$\{NormalWord\}\(\\\-$\{NormalWord\}\)\*?$/;

print "ok 1\\n" if $a =~ $PredNameHyphen;

$a =~ /^$/;

print "ok 2\\n" if $a =~ $PredNameHyphen;

1;

\_\_END\_\_

\#    output&#8203;:

Perl v5\.6\.0
a ='x\-x'
au='x\-x'
ok 1
ok 2
\#    ok

Perl v5\.6\.0
a ='x\-ä'
au='x\-Ã¤'
ok 1
ok 2
\#    ok

Perl 5\.8\.0&#8203;:
a ='x\-x'
au='x\-x'
ok 1
\#    fails at match 2

Perl 5\.8\.0&#8203;:
a ='x\-ä'
au='x\-Ã¤'
\#    fails at both matches \!

\-\-\-\- CUT \-\-\-\-


\-\- 
&nbsp;              ProgramV \- Alice on Perl \- available at
&nbsp;               http&#8203;://www\.virtualitas\.net/perl/aiml/

&nbsp;      VIRTUALITAS \- Manufacturer of fine OOPPS \- since 1996
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\* VIRTUALITAS Inc\.               \*      http&#8203;://www\.virtualitas\.net  \*
\*                                \*                                  \*
\* European Consultant Office     \*                                  \*
\* Internationales Handelszentrum \*                                  \*
\* Friedrichstraße 95             \*   contact&#8203;:Ernest Lergon          \*
\* 10117 Berlin / Germany         \*    mailto&#8203;:Ernest@&#8203;virtualitas\.net \*
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
&nbsp;      PGP\-Fingerprint 6E6F DC17 A886 342D  D63F 7880 12F5 6BA9
&nbsp;        PGP\-Key http&#8203;://www\.virtualitas\.net/Ernest\_Lergon\.asc

**Comment by p5pRT** [comment]
### From @jhi

This bug was recently fixed \(changes \#18653 and \#18658\, for those following
the perl5\-changes mailing list\)\.  I'm adding the test case from this problem ticket
just to be paranoid\, and then marking the problem ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
